### ISSUE TITLE
Restrictive permissions on non-root sandbox - build failures / cannot delete

### ISSUE THREAD

**Issue by dtrudg** [apptainer/singularity#4524]
**This is one issue extracted from #4517** reported originaly by @afortiorama

### Version of Singularity:

3.4.0 or greater

### Expected behavior

When building a `docker://centos:7` sandbox as a non-root user, the build completes. When finished with the sandbox, it can be removed without issue by the non-root user.

### Actual behavior

There are 2 symptoms...

(1) *IF* the sandbox final location (e.g. `$HOME/test_centos`) is *not* on the same filesystem as `TMPDIR` then the build will fail...

```
dave@piran~> singularity build --sandbox test_centos docker://centos:7
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:d8d02d45731499028db01b6fa35475f91d230628b4e25fab8e3c015594dc3261
Copying config sha256:acab94af64effb1f7481666a37788e7a59465e723f0b0fe0a0f458f3f4856638
 1.05 KiB / 1.05 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
2019/09/25 19:47:47  info unpack layer: sha256:d8d02d45731499028db01b6fa35475f91d230628b4e25fab8e3c015594dc3261
2019/09/25 19:47:47  warn rootless{usr/bin/ping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2019/09/25 19:47:48  warn rootless{usr/sbin/arping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2019/09/25 19:47:48  warn rootless{usr/sbin/clockdiff} ignoring (usually) harmless EPERM on setxattr "security.capability"
INFO:    Creating sandbox directory...
FATAL:   While performing build: sandbox assemble failed: exit status 1: mv: cannot open '/tmp/sbuild-474633384/fs/etc/gshadow' for reading: Permission denied
mv: cannot open '/tmp/sbuild-474633384/fs/etc/gshadow-' for reading: Permission denied
mv: cannot open '/tmp/sbuild-474633384/fs/etc/shadow' for reading: Permission denied
mv: cannot open '/tmp/sbuild-474633384/fs/etc/shadow-' for reading: Permission denied
```

(2) If the sandbox is built to the same filesystem as `TMPDIR` it will build okay, but it is not possible to delete it due to there being files/directories present with permissions that do not allow deletion:

```
$ singularity build --sandbox test_samefs docker://centos:7
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:d8d02d45731499028db01b6fa35475f91d230628b4e25fab8e3c015594dc3261
Copying config sha256:acab94af64effb1f7481666a37788e7a59465e723f0b0fe0a0f458f3f4856638
 1.05 KiB / 1.05 KiB [======================================================] 0s
Writing manifest to image destination
Storing signatures
2019/09/25 19:52:10  info unpack layer: sha256:d8d02d45731499028db01b6fa35475f91d230628b4e25fab8e3c015594dc3261
2019/09/25 19:52:10  warn rootless{usr/bin/ping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2019/09/25 19:52:11  warn rootless{usr/sbin/arping} ignoring (usually) harmless EPERM on setxattr "security.capability"
2019/09/25 19:52:11  warn rootless{usr/sbin/clockdiff} ignoring (usually) harmless EPERM on setxattr "security.capability"
INFO:    Creating sandbox directory...
INFO:    Build complete: test_samefs

$ rm -rf test_samefs
rm: cannot remove 'test_samefs/root/.bash_profile': Permission denied
rm: cannot remove 'test_samefs/root/.bash_logout': Permission denied
rm: cannot remove 'test_samefs/root/.cshrc': Permission denied
rm: cannot remove 'test_samefs/root/.bashrc': Permission denied
rm: cannot remove 'test_samefs/root/.tcshrc': Permission denied
rm: cannot remove 'test_samefs/root/anaconda-ks.cfg': Permission denied
rm: cannot remove 'test_samefs/usr/share/licenses/device-mapper-libs-1.02.149/COPYING': Permission denied
rm: cannot remove 'test_samefs/usr/share/licenses/device-mapper-libs-1.02.149/COPYING.LIB': Permission denied
rm: cannot remove 'test_samefs/usr/bin/sqlite3': Permission denied
rm: cannot remove 'test_samefs/usr/bin/comm': Permission denied
rm: cannot remove 'test_samefs/usr/bin/testgdbm': Permission denied
rm: cannot remove 'test_samefs/usr/bin/ln': Permission denied
rm: cannot remove 'test_samefs/usr/bin/show-installed': Permission denied
rm: cannot remove 'test_samefs/usr/bin/lslocks': Permission denied
rm: cannot remove 'test_samefs/usr/bin/grep': Permission denied
rm: cannot remove 'test_samefs/usr/bin/ipcmk': Permission denied
rm: cannot remove 'test_samefs/usr/bin/repoclosure': Permission denied
rm: cannot remove 'test_samefs/usr/bin/hostid': Permission denied
rm: cannot remove 'test_samefs/usr/bin/glib-compile-schemas': Permission denied
...
```

This is caused by the change to `umoci` for unpacking OCI/docker layers into a rootfs. We previously used a patched `image-tools` for extraction that ensured minimum permissions on files / dirs so that they could be moved (tmp location -> sandbox location) or deleted by a non-root user.

See the `image-tools` patch here: https://github.com/sylabs/image-tools/commit/2814f498056809a9d5baaf76d1d82312180a5888

A similar thing was done in the `docker-extract` libarchive C routines in singularity `2.x` for the same reason.

**Comment by dtrudg** [comment]
This issue is currently being tackled via work ongoing in #4522 - to restore the permissions of OCI layers on rootless extraction as they were prior to Singularity 3.4.0 and the `image-tools` -> `umoci` extraction changes

**Comment by dtrudg** [comment]
A minimal docker image to reproduce this can be found here: https://cloud.docker.com/u/sylabsio/repository/docker/sylabsio/issue4524

On a system where my `$HOME` is not on same fs as `TMPDIR`...

```
08:34 PM $ singularity build --sandbox docker4524 docker://sylabsio/issue4524
INFO:    Starting build...
Getting image source signatures
Skipping fetch of repeat blob sha256:7c9d20b9b6cda1c58bc4f9d6c401386786f584437abbe87e58910f8a9a15386b
Skipping fetch of repeat blob sha256:4505ad60a1aa21f42c0584e037044d2195b3e5c1a718b5f7fe5c94eb3b788c46
Copying config sha256:19b94cc8a74f4c20c7dd3b1109988337af7e741c5873338a732140ae934e5b6a
 798 B / 798 B [============================================================] 0s
Writing manifest to image destination
Storing signatures
2019/09/25 20:34:14  info unpack layer: sha256:7c9d20b9b6cda1c58bc4f9d6c401386786f584437abbe87e58910f8a9a15386b
2019/09/25 20:34:14  info unpack layer: sha256:4505ad60a1aa21f42c0584e037044d2195b3e5c1a718b5f7fe5c94eb3b788c46
INFO:    Creating sandbox directory...
FATAL:   While performing build: sandbox assemble failed: exit status 1: mv: cannot access '/tmp/sbuild-760517915/fs/directory': Permission denied
mv: cannot open '/tmp/sbuild-760517915/fs/file' for reading: Permission denied
```

The dockerfile for this is:

```
FROM busybox
RUN mkdir directory && touch file && chmod 000 directory file
```

**Comment by dtrudg** [comment]
This is closed in 3.4.2 via #4522. A related issue for master remains in #4593 that needs to be attacked in a more targeted manner.

**Comment by dtrudg** [comment]
This non-removable sandbox component of this issue is addressed on `master` for singularity 3.5 and on via the `--fix-perms` flag implemented, and tested in #4672
